JSon data structures don't own their strings

This allows zero-copy json parsing, at the cost of complexifying the lifetime handling, and not being able to parse from a stream (as the json string from which the string slices are borrowed must outlive the deserialized structure), thus making Bundle::from_file currently not buildable.

I guess we must choose either to
- support loading from a stream (thus making the Bundle and substructures own their strings)
- only support the from_string with constraints on the source string lifetime.

This was introduced by #4. Here's the error that I get:

```
 cargo test
   Compiling libcnab v0.1.0 (/home/technosophos/Code/Rust/libcnab)
error[E0308]: mismatched types
  --> src/cnab.rs:68:27
   |
68 |         let res: Bundle = serde_json::from_reader(file)?;
   |                           ^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `serde::Deserialize<'de>`
              found type `serde::Deserialize<'_>`

error[E0308]: mismatched types
  --> src/cnab.rs:68:27
   |
68 |         let res: Bundle = serde_json::from_reader(file)?;
   |                           ^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `serde::Deserialize<'de>`
              found type `serde::Deserialize<'_>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `libcnab`.
warning: build failed, waiting for other jobs to finish...
error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `libcnab`.

To learn more, run the command again with --verbose.
```

I haven't figured out a work-around for the error above that doesn't involve reading the file into a string and calling the string parser.

--- a/src/cnab.rs
+++ b/src/cnab.rs
@@ -11,54 +11,54 @@ use std::path::Path;
 /// The fields here are in canonical order.
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct Bundle<'a> {
+pub struct Bundle {
     /// The list of additional actions that this bundle can perform.
     ///
     /// 'install', 'upgrade', and 'uninstall' are default actions, but additional actions
     /// may be defined here.
-    pub actions: Option<HashMap<&'a str, Action<'a>>>,
+    pub actions: Option<HashMap<String, Action>>,
     /// The list of configurable credentials.
     ///
     /// Credentials are injected into the bundle's invocation image at startup time.
-    pub credentials: Option<Vec<Credential<'a>>>,
+    pub credentials: Option<HashMap<String, Credential>>,
     /// This field allows for additional data to described in the bundle.
     ///
     /// This data should be stored in key/value pairs, where the value is undefined by
     /// the specification (but must be representable as JSON).
-    pub custom: Option<HashMap<&'a str, serde_json::Value>>,
+    pub custom: Option<HashMap<String, serde_json::Value>>,
     /// description is a short description of this bundle
-    pub description: Option<&'a str>,
+    pub description: Option<String>,
     /// The list of images that comprise this bundle.
     ///
     /// Each image here is considered a constituent of the application described by this
     /// bundle.
-    pub images: Option<HashMap<&'a str, Image<'a>>>,
+    pub images: Option<HashMap<String, Image>>,
     /// inovcation_images is the list of available bootstrapping images for this bundle
     ///
     /// Only one ought to be executed.
-    pub invocation_images: Vec<Image<'a>>,
+    pub invocation_images: Vec<Image>,
     /// keywords is a list of keywords describing this bundle
-    pub keywords: Option<Vec<&'a str>>,
+    pub keywords: Option<Vec<String>>,
     /// license is the license of this bundle
-    pub license: Option<&'a str>,
+    pub license: Option<String>,
     /// maintainers is a list of maintainers responsible for this bundle
-    pub maintainers: Option<Vec<Maintainer<'a>>>,
+    pub maintainers: Option<Vec<Maintainer>>,
     /// name is the name of the bundle
-    pub name: &'a str,
+    pub name: String,
     /// The collection of parameters that can be passed into this bundle.
     ///
     /// Parameters can be injected into a bundle during startup time.
-    pub parameters: Option<HashMap<&'a str, Parameter<'a>>>,
+    pub parameters: Option<HashMap<String, Parameter>>,
     /// schema_version is the version of the CNAB specification used to describe this
-    pub schema_version: &'a str,
+    pub schema_version: String,
     /// version is the version of the bundle
-    pub version: &'a str,
+    pub version: String,
 }
 
 /// Represents a bundle.
-impl<'a> Bundle<'a> {
+impl Bundle {
     ///fn new(name: String, version: String) -> Bundle {}
-    pub fn from_string(json_data: &'a str) -> Result<Bundle<'a>, serde_json::Error> {
+    pub fn from_string(json_data: &str) -> Result<Bundle, serde_json::Error> {
         let res: Bundle = serde_json::from_str(json_data)?;
         Ok(res)
     }
@@ -74,13 +74,13 @@ impl<'a> Bundle<'a> {
 ///
 /// The name field is required, though the format of its value is unspecified.
 #[derive(Debug, Serialize, Deserialize)]
-pub struct Maintainer<'a> {
+pub struct Maintainer {
     /// The email address of the maintainer
-    pub email: Option<&'a str>,
+    pub email: Option<String>,
     /// The name of the maintainer
-    pub name: &'a str,
+    pub name: String,
     /// A URL with more information about the maintainer
-    pub url: Option<&'a str>,
+    pub url: Option<String>,
 }
 
 /// Image describes a CNAB image.
@@ -88,43 +88,43 @@ pub struct Maintainer<'a> {
 /// Both invocation images and regular images can be described using this object.
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct Image<'a> {
+pub struct Image {
     /// A digest to be used to verify the integrity of the image
-    pub digest: Option<&'a str>,
+    pub digest: Option<String>,
     /// The image, as a string of the form REPO/NAME:TAG@SHA
-    pub image: &'a str,
+    pub image: String,
     /// The type of image. Typically, this is treated as an OCI Image
-    pub image_type: Option<&'a str>,
+    pub image_type: Option<String>,
     /// The media type of the image
-    pub media_type: Option<&'a str>,
+    pub media_type: Option<String>,
     /// The platform this image may be deployed on
-    pub platform: Option<Platform<'a>>,
+    pub platform: Option<Platform>,
     /// The size in bytes of the image
     pub size: Option<i64>,
 }
 
 /// Platform defines a platform as a machine architecture plus and operating system
 #[derive(Debug, Serialize, Deserialize)]
-pub struct Platform<'a> {
+pub struct Platform {
     /// The architecture
     ///
     /// Typical values are amd64, i386, and arm64
-    pub arch: Option<&'a str>,
+    pub arch: Option<String>,
     /// The operating system.
     ///
     /// Typical values are darwin, windows, and linux
-    pub os: Option<&'a str>,
+    pub os: Option<String>,
 }
 
 /// Credential describes a particular credential that may be injected into a bundle
 #[derive(Debug, Serialize, Deserialize)]
-pub struct Credential<'a> {
+pub struct Credential {
     /// The description of this credential
-    pub description: Option<&'a str>,
+    pub description: Option<String>,
     /// The name of the environment variable into which the value will be placed
-    pub env: Option<&'a str>,
+    pub env: Option<String>,
     /// The fully qualified path into which the value will be placed
-    pub path: Option<&'a str>,
+    pub path: Option<String>,
 }
 
 /// Parameter describes a parameter that will be put into the invocation image
@@ -132,13 +132,13 @@ pub struct Credential<'a> {
 /// Paramters are injected into the invocation image at startup time
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct Parameter<'a> {
+pub struct Parameter {
     /// The actions to which this parameter applies.
     ///
     /// If unset, this parameter will be applied to all actions.
-    pub apply_to: Option<Vec<&'a str>>,
+    pub apply_to: Option<Vec<String>>,
     /// The location where this parameter will be injected in the invocation image
-    pub destination: Destination<'a>,
+    pub destination: Destination,
     /// This parameter's default value
     pub default_value: Option<serde_json::Value>,
 
@@ -163,7 +163,7 @@ pub struct Parameter<'a> {
     /// If unspecified, no max is applied.
     pub max_length: Option<i64>,
     /// Additional parameter information
-    pub metadata: Option<Metadata<'a>>,
+    pub metadata: Option<Metadata>,
     /// The minimum integer value
     ///
     /// If unspecified, the minimum 64-bit integer value is applied
@@ -173,7 +173,7 @@ pub struct Parameter<'a> {
     /// A regular expression (as defined in ECMAScript)
     ///
     /// If it is not matched, a string parameter value will be rejected
-    pub pattern: Option<&'a str>,
+    pub pattern: Option<String>,
     /// Indicate whether this parameter is required
     ///
     /// Default is false.
@@ -181,7 +181,7 @@ pub struct Parameter<'a> {
     pub required: bool,
     /// This describes the underlying type of the parameter (string, int...)
     #[serde(rename = "type")]
-    pub parameter_type: &'a str, // Should be Enum; alphabetically, this is 'type'
+    pub parameter_type: String, // Should be Enum; alphabetically, this is 'type'
 }
 
 /// An Action is a custom action in an invocation image.
@@ -189,9 +189,9 @@ pub struct Parameter<'a> {
 /// For example, an invocation image may provide help text by creating a 'help'
 /// action that, when triggered, prints help text to STDOUT.
 #[derive(Debug, Serialize, Deserialize)]
-pub struct Action<'a> {
+pub struct Action {
     /// Describes what this action does
-    pub description: Option<&'a str>,
+    pub description: Option<String>,
     /// If true, this action modifies the deployment, and should be tracked as a release.
     #[serde(default)]
     pub modifies: bool,
@@ -205,9 +205,9 @@ pub struct Action<'a> {
 
 /// Describe a parameter
 #[derive(Debug, Serialize, Deserialize)]
-pub struct Metadata<'a> {
+pub struct Metadata {
     /// A description of a parameter
-    pub description: Option<&'a str>,
+    pub description: Option<String>,
 }
 
 /// Destination describes where, in the invocation image, a particular paramter value should be
@@ -217,9 +217,9 @@ pub struct Metadata<'a> {
 /// a particular location on the filesystem (`path`). This is a non-exclusive or, meaining
 /// that the same paramter can be written to both an env var and a path.
 #[derive(Debug, Serialize, Deserialize)]
-pub struct Destination<'a> {
+pub struct Destination {
     /// The name of the destination environment variable
-    pub env: Option<&'a str>,
+    pub env: Option<String>,
     /// The fully qualified path to the destination file
-    pub path: Option<&'a str>,
+    pub path: Option<String>,
 }

--- a/src/tests.rs
+++ b/src/tests.rs
@@ -18,9 +18,9 @@ fn test_bundle_simple() {
 
     let bun = res.unwrap();
 
-    assert_that(&bun.name).is_equal_to("aristotle");
-    assert_that(&bun.schema_version).is_equal_to("1.0-WD");
-    assert_that(&bun.version).is_equal_to("1.0.0");
+    assert_that(&bun.name).is_equal_to("aristotle".to_string());
+    assert_that(&bun.schema_version).is_equal_to("1.0-WD".to_string());
+    assert_that(&bun.version).is_equal_to("1.0.0".to_string());
     assert_that(&bun.invocation_images.len()).is_equal_to(&0);
 }
 
@@ -39,16 +39,16 @@ fn test_bundle_keywords() {
 
     let bun = res.unwrap();
 
-    assert_that(&bun.name).is_equal_to("aristotle");
-    assert_that(&bun.schema_version).is_equal_to("1.0-WD");
-    assert_that(&bun.version).is_equal_to("1.0.0");
+    assert_that(&bun.name).is_equal_to("aristotle".to_string());
+    assert_that(&bun.schema_version).is_equal_to("1.0-WD".to_string());
+    assert_that(&bun.version).is_equal_to("1.0.0".to_string());
     assert_that(&bun.invocation_images.len()).is_equal_to(&0);
 
     let kw = &bun.keywords.unwrap();
     assert_that(&kw.len()).is_equal_to(&3);
-    assert_that(&kw[0]).is_equal_to("a");
-    assert_that(&kw[1]).is_equal_to("b");
-    assert_that(&kw[2]).is_equal_to("c");
+    assert_that(&kw[0]).is_equal_to("a".to_string());
+    assert_that(&kw[1]).is_equal_to("b".to_string());
+    assert_that(&kw[2]).is_equal_to("c".to_string());
 }
 
 #[test]
@@ -108,27 +108,27 @@ fn test_bundle_parameters() {
 
     let bun = res.unwrap();
 
-    assert_that(&bun.name).is_equal_to("aristotle");
-    assert_that(&bun.schema_version).is_equal_to("1.0-WD");
-    assert_that(&bun.version).is_equal_to("1.0.0");
+    assert_that(&bun.name).is_equal_to("aristotle".to_string());
+    assert_that(&bun.schema_version).is_equal_to("1.0-WD".to_string());
+    assert_that(&bun.version).is_equal_to("1.0.0".to_string());
 
     let params = bun.parameters.unwrap();
     assert_that(&params.len()).is_equal_to(&3);
 
-    let arg3 = params.get(&"arg3");
+    let arg3 = params.get(&"arg3".to_string());
 
     assert!(arg3.is_some());
-    assert_that(&arg3.unwrap().parameter_type).is_equal_to("string");
+    assert_that(&arg3.unwrap().parameter_type).is_equal_to("string".to_string());
 
     let apply = &arg3.unwrap().apply_to;
     assert!(apply.is_some());
 
     let dest = &arg3.unwrap().destination;
     let env = &dest.env;
-    assert_that(&env).is_equal_to(&Some("LETTERS"));
+    assert_that(&env).is_equal_to(&Some("LETTERS".to_string()));
 
     let path = &dest.path;
-    assert_that(path).is_equal_to(&Some("/path/to/abc"));
+    assert_that(path).is_equal_to(&Some("/path/to/abc".to_string()));
 
     let abc = json!("abc");
     let dv = &arg3.unwrap().default_value;
@@ -139,17 +139,17 @@ fn test_bundle_parameters() {
 
     assert_that(&arg3.as_ref().unwrap().min_length.unwrap()).is_equal_to(1);
     assert_that(&arg3.as_ref().unwrap().max_length.unwrap()).is_equal_to(5);
-    assert_that(&arg3.as_ref().unwrap().pattern).is_equal_to(&Some("[a-z]+"));
+    assert_that(&arg3.as_ref().unwrap().pattern).is_equal_to(&Some("[a-z]+".to_string()));
     assert_that(&arg3.unwrap().required).is_equal_to(true);
 
     let meta = &arg3.unwrap().metadata;
     assert_that(&meta.is_some()).is_equal_to(true);
 
-    assert_that(&meta.as_ref().unwrap().description.as_ref().unwrap()).is_equal_to(&"a parameter");
+    assert_that(&meta.as_ref().unwrap().description.as_ref().unwrap()).is_equal_to(&"a parameter".to_string());
 
     let apply_to = &arg3.unwrap().apply_to;
-    assert_that(apply_to).is_equal_to(&Some(vec!["uninstall"]));
-    assert_that(&arg3.unwrap().parameter_type).is_equal_to("string");
+    assert_that(apply_to).is_equal_to(&Some(vec!["uninstall".to_string()]));
+    assert_that(&arg3.unwrap().parameter_type).is_equal_to("string".to_string());
 }
 
 #[test]
@@ -190,9 +190,9 @@ fn test_bundle_images() {
 
     let bun = res.unwrap();
 
-    assert_that(&bun.name).is_equal_to("aristotle");
-    assert_that(&bun.schema_version).is_equal_to("1.0-WD");
-    assert_that(&bun.version).is_equal_to("1.0.0");
+    assert_that(&bun.name).is_equal_to("aristotle".to_string());
+    assert_that(&bun.schema_version).is_equal_to("1.0-WD".to_string());
+    assert_that(&bun.version).is_equal_to("1.0.0".to_string());
     assert_that(&bun.invocation_images.len()).is_equal_to(&1);
 }
 

